L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,425 @@
+  // |watch_identity_key_cert_pairs| is called. If not set, an empty string will
+  // be used as the name.
+  //
+  // @param identity_cert_name the name of identity key-certificate pairs being set.
+  void set_identity_cert_name(const std::string& identity_cert_name);
+
+  // Sets the TLS session key logging file path. If not set, TLS
+  // session key logging is disabled. Note that this should be used only for
+  // debugging purposes. It should never be used in a production environment
+  // due to security concerns.
+  //
+  // @param tls_session_key_log_file_path: Path where TLS session keys should
+  // be logged.
+  void set_tls_session_key_log_file_path(
+      const std::string& tls_session_key_log_file_path);
+
+  // Sets the certificate verifier. The certificate verifier performs checks on
+  // the peer certificate chain after the chain has been (cryptographically)
+  // verified to chain up to a trusted root.
+  void set_certificate_verifier(
+      std::shared_ptr<CertificateVerifier> certificate_verifier);
+
+  // Sets the crl provider, see CrlProvider for more details.
+  void set_crl_provider(std::shared_ptr<CrlProvider> crl_provider);
+
+  // Sets the minimum TLS version that will be negotiated during the TLS
+  // handshake. If not set, the underlying SSL library will default to TLS v1.2.
+  // @param tls_version: The minimum TLS version.
+  void set_min_tls_version(grpc_tls_version tls_version);
+
+  // Sets the maximum TLS version that will be negotiated during the TLS
+  // handshake. If not set, the underlying SSL library will default to TLS v1.3.
+  // @param tls_version: The maximum TLS version.
+  void set_max_tls_version(grpc_tls_version tls_version);
+
+  // Sets a custom chain builder implementation that replaces the default chain
+  // building from the underlying SSL library. Fully replacing and implementing
+  // chain building is a complex task and has dangerous security implications if
+  // done wrong, thus this API is inteded for expert use only.
+  void set_custom_chain_builder(std::shared_ptr<CustomChainBuilder>
+  chain_builder);
+}
+
+// Server-specific options for configuring TLS.
+class TlsServerCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // A certificate provider that provides identity credentials is required,
+  // because a server must always present identity credentials during any TLS
+  // handshake. The certificate provider may optionally provide root
+  // certificates, in case the server requests client certificates.
+  explicit TlsServerCredentialsOptions(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider)
+      : TlsCredentialsOptions() {
+    set_certificate_provider(certificate_provider);
+  }
+
+
+  // Sets requirements for if client certificates are requested, if they
+  // are required, and if the client certificate must be trusted.  The
+  // default is GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE.
```

Comment:
Maybe we should explicitly say that the default corresponds to doing "normal TLS"?

Reply:
Done